### PR TITLE
git-sync --all

### PR DESCRIPTION
**Pull request by charlierudolph**
@kevgo @allewun 

I broke up first and last branch in all conflict cases to make sure aborting / skipping / continuing worked as expected.

Tests are structured as follows:

* feature_branch_merge_main_branch_conflict
  * with_tracking_branch_updates
    * first_branch
       * with_open_changes.feature
       * without_open_changes.feature
    * last_branch
       * with_open_changes.feature
       * without_open_changes.feature
  * without_tracking_branch_updates
    * first_branch
       * with_open_changes.feature
       * without_open_changes.feature
    * last_branch
       * with_open_changes.feature
       * without_open_changes.feature
* feature_branch_merge_tracking_branch_conflict
    * first_branch
       * with_open_changes.feature
       * without_open_changes.feature
    * last_branch
       * with_open_changes.feature
       * without_open_changes.feature
* main_branch_rebase_tracking_branch_conflict
  * with_open_changes.feature
  * without_open_changes.feature
* no_conflicts
   * feature_branches
     * with_open_changes.feature
     * without_open_changes.feature
   * non_feature_branches
     * with_open_changes.feature
     * without_open_changes.feature
   * remote_only_branches
     * with_open_changes.feature
     * without_open_changes.feature
* non_feature_branch_rebase_tracking_branch_conflict
  * first_branch
     * with_open_changes.feature
     * without_open_changes.feature
  * last_branch
     * with_open_changes.feature
     * without_open_changes.feature

### CODE REVIEW

**Comment by allewun** [comment]
I like git sync --all over git sync-all

**Comment by kevgo** [comment]
+1 on `git sync --all`

File: features/git-sync-all/feature_branch_merge_main_branch_conflict/with_tracking_branch_updates/first_branch/with_open_changes.feature [hunk 1]
```diff
@@ -0,0 +1,135 @@
+Feature: git-sync-all: handling merge conflicts between feature branch and main branch with open changes
+
+  Background:
+    Given I have feature branches named "feature1" and "feature2"
+    And the following commits exist in my repository
+      | branch   | location         | message                | file name            | file content            |
```
Comment (kevgo) [inline features/git-sync-all/feature_branch_merge_main_branch_conflict/with_tracking_branch_updates/first_branch/with_open_changes.feature]: Upcase the table headers

File: features/git-sync-all/feature_branch_merge_main_branch_conflict/with_tracking_branch_updates/first_branch/with_open_changes.feature [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      |          |                  | main commit                                                  | conflicting_file     |
+      |          |                  | Merge remote-tracking branch 'origin/feature1' into feature1 |                      |
+      |          |                  | feature1 local commit                                        | conflicting_file     |
+      |          |                  | feature1 remote commit                                       | feature1_remote_file |
+      | feature2 | local and remote | Merge branch 'main' into feature2                            |                      |
+      |          |                  | main commit                                                  | conflicting_file     |
+      |          |                  | feature2 commit                                              | feature2_file        |
+
+
+  Scenario: continuing after resolving conflicts and committing
```
Comment (kevgo) [inline features/git-sync-all/feature_branch_merge_main_branch_conflict/with_tracking_branch_updates/first_branch/with_open_changes.feature]: In other specs we also tested the scenario to continue without committing here. If you only want to check one, I think that one is a bit more inclusive than the one with committing.

File: features/git-sync-all/feature_branch_merge_main_branch_conflict/with_tracking_branch_updates/last_branch/with_open_changes.feature [hunk 3]
```diff
@@ -0,0 +1,130 @@
+Feature: git-sync-all: handling merge conflicts between feature branch and main branch with open changes
+
+  Background:
+    Given I have feature branches named "feature1" and "feature2"
+    And the following commits exist in my repository
+      | branch   | location         | message                | file name            | file content            |
```
Comment (kevgo) [inline features/git-sync-all/feature_branch_merge_main_branch_conflict/with_tracking_branch_updates/last_branch/with_open_changes.feature]: Upcase table headers everywhere

File: features/git-sync-all/no_conflicts/feature_branches/without_open_changes.feature [hunk 4]
```diff
@@ -0,0 +1,39 @@
+Feature: git-sync-all: syncs all feature branches with open changes
```
Comment (kevgo) [inline features/git-sync-all/no_conflicts/feature_branches/without_open_changes.feature]: You flipped the "with/without open changes" description. This spec is without open changes, but internally says it is with. Same for the other file.

File: src/git-sync-all [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  fetch
+}
+
+
+function skip_message {
+  echo "To skip the sync of the '$(get_current_branch_name)' branch"
+}
+
+
+function skippable {
```
Comment (kevgo) [inline src/git-sync-all]: `# Returns whether synchronization of the current branch can be skipped by the user`

File: src/git-sync-all [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+source "$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )/helpers/helpers.sh"
+
+
+function preconditions {
+  fetch
+}
+
+
+function skip_message {
+  echo "To skip the sync of the '$(get_current_branch_name)' branch"
```
Comment (kevgo) [inline src/git-sync-all]: This is a grammatically strange output. Please either improve it, or add a comment describing what exactly this method is for.

File: src/helpers/script_helpers.sh [hunk 7]
```diff
@@ -122,3 +145,42 @@ function run_steps {
...
+
+      break
+    else
+      remove_line "$file"
+    fi
+  done
+}
+
+
+# Placeholder for any scripts that do have the skip interface
```
Comment (kevgo) [inline src/helpers/script_helpers.sh]: `# Skipping must be explicitly allowed by the respective script.`  ?

File: src/helpers/script_helpers.sh [hunk 8]
```diff
@@ -122,3 +145,42 @@ function run_steps {
...
+function skip_current_branch_steps {
+  local file="$1"
+  local add_noop="$2"
+
+  while [ "$(has_lines "$file")" = true ]; do
+    if [[ "$(peek_line "$file")" =~ ^checkout ]]; then
+      if [ -n "$add_noop" ]; then
+        prepend_to_file "noop" "$file"
+      fi
+
```
Comment (kevgo) [inline src/helpers/script_helpers.sh]: I don't think this empty line adds readability. I find it rather confusing, and would vote to remove it.

**Comment by kevgo** [comment]
This is a real masterpiece, @charlierudolph! Very well done. I'll look over this again in the morning, but this looks pretty shippable to me already!

**Comment by kevgo** [comment]
We should document better why there are sometimes no-ops now. This isn't fully obvious from quickly glancing at the code. A few months in the future, when you and I have forgotten why this had to be done, we will be grateful to have those comments to get back up to speed on this issue.

**Comment by charlierudolph** [comment]
I have a comment about the "noop" in "push_helpers.rb" where I actually call "skip_current_branch_steps" with "add_noop".

**Comment by kevgo** [comment]
Yes, I saw that. It was a bit thin. Can you explain a bit more why the no-op is necessary there?

File: src/git-sync-all (hunk 5 above)
Comment (charlierudolph) [inline src/git-sync-all]: This is more generic. skippable is a function used by the command list architecture and is just used to determine whether or not to print the a message saying you can run `--skip`. I will add a comment to the placeholder function in script_helpers.

File: src/git-sync-all (hunk 6 above)
Comment (charlierudolph) [inline src/git-sync-all]: Its only the first part of a sentence. What about renaming the function to skip_message_prefix?

Comment (kevgo) [inline src/git-sync-all]: That would make clear enough that this is only a prefix.

File: src/git-sync-all (hunk 5 above)
Comment (kevgo) [inline src/git-sync-all]: Okay

**Comment by charlierudolph** [comment]
@kevgo all comments addressed and converted to `git sync --all`

**Comment by kevgo** [comment]
:guitar:   (this is an LGTM for commits that totally rock)

**Pull request closed without merging**